src/graphnet/models/task/reconstruction.py

```diff
@@ -167,6 +167,13 @@ def _forward(self, x):
         # Leave as it is
         return x
 
+class MulticlassificationTask(Task):
+    # requires three features: probability of being noise, muon or neutrino.
+    nb_inputs = 3
+
+    def _forward(self, x):
+        # transform probability of being noise, muon or neutrino
+        return torch.sigmoid(x)
```

Comment:
This hard-codes the `Task` to three-way classification, which does not align with the class name. I think we need to do something a bit more dynamical, see (untested) sketched suggestion below. You also need to use softmax activation for multi-class classification to ensure that outputs are in [0,1] and sum to one. (The last condition is not satisfied by sigmoid.)
```suggestion
class MulticlassClassificationTask(Task):
    # Requires the same number of features as the number of classes being predicted
    @property
    def nb_inputs(self):
        return self._nb_inputs

    @save_config
    def __init__(self, nb_inputs, *args, **kwargs):
        self._nb_inputs = nb_inputs
        self._softmax = torch.nn.Softmax()
        super().__init__(self, *args, **kwargs)

    def _forward(self, x):
        # Transform latent features into probabilities.
        return self._softmax(x)
```

Reply:
This is true. The predictions that I ran gave results that did not add up to 1, i.e. 
predictions of {noise, muon, neutrino} = {7.072934220832261e-16, 5.221908991911821e-14, 1.0}

Reply:
can you elaborate on the `@save_config` decorator?
and would the implementation of the `__init__()` imply that the task would have two additional callables? i.e.
```
task = MulticlassClassificationTask(
        nb_inputs=?,
        softmax= <default torch.nn.Softmax(),
        hidden_size=gnn.nb_outputs,
        target_labels=config["target"],
        loss_function=CrossEntropyLoss(),
    )
```

Reply:
When Softmax() is called in:
```
def __init__(self, nb_inputs, *args, **kwargs):
    self._nb_inputs = nb_inputs
    self._softmax = torch.nn.Softmax()
    super().__init__(self, *args, **kwargs)
```
The following error occour
```
Exception has occurred: AttributeError
cannot assign module before Module.__init__() call
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/task/reconstruction.py", line 177, in __init__
    self._softmax = torch.nn.Softmax()
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 197, in main
    task = MulticlassClassificationTask(
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 261, in <module>
    main()
```
[Consulting stack](https://stackoverflow.com/a/43080779), it should be a matter of calling the super().__init__() correctly, but I haven't been able to construct it in such a way that would fix the issue.

Reply:
Try moving the `super().__init__(...)` call up as the first statement in the constructor. 🙂

Reply:
> Try moving the `super().__init__(...)` call up as the first statement in the constructor. 🙂

I tried to reorder it, but then It complains about `hidden_size`
`Argument of type "Self@MulticlassClassificationTask" cannot be assigned to parameter "hidden_size" of type "int" in function "__init__"`
It gives approximately the same error in the traceback when I try to run it.
`hidden_size` can also not be assigned as a `self.` statement in the init

Reply:
> This is true. The predictions that I ran gave results that did not add up to 1, i.e. predictions of {noise, muon, neutrino} = {7.072934220832261e-16, 5.221908991911821e-14, 1.0}

I think this would still qualify as summing to 1, since the first two numbers are probably below `eps` for the type of float you're using. But the softmax point should still apply. :)

Reply:
> > Try moving the `super().__init__(...)` call up as the first statement in the constructor. 🙂
> 
> I tried to reorder it, but then It complains about `hidden_size` `Argument of type "Self@MulticlassClassificationTask" cannot be assigned to parameter "hidden_size" of type "int" in function "__init__"` It gives approximately the same error in the traceback when I try to run it. `hidden_size` can also not be assigned as a `self.` statement in the init

Try 
```
def __init__(self, nb_inputs, *args, **kwargs):
    self._nb_inputs = nb_inputs
    super().__init__(self, *args, **kwargs)
    self._softmax = torch.nn.Softmax()
```
Otherwise, could you paste, or point to, the current version of the relevant code snippet along with the full error message?

Reply:
I am somewhat comforted in you trying all the same things I've already tried :D
```
def __init__(self, nb_inputs, *args, **kwargs):
    super().__init__(self, *args, **kwargs)
    self._nb_inputs = nb_inputs
    self._softmax = softmax()
```
The above code has the following trace:
```
Exception has occurred: TypeError
__init__() got multiple values for argument 'hidden_size'
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/task/reconstruction.py", line 178, in __init__
    super().__init__(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 206, in main
    task = MulticlassClassificationTask(
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 270, in <module>
    main()
```
```
def __init__(self, nb_inputs, *args, **kwargs):
    super(Task, self).__init__(self, *args, **kwargs)
    self._nb_inputs = nb_inputs
    self._softmax = softmax()
```
The above code initially removes the hidden_size message in pylance but results in the following trace
```
Exception has occurred: TypeError
__init__() got an unexpected keyword argument 'hidden_size'
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/task/reconstruction.py", line 178, in __init__
    super(Task,self).__init__(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 206, in main
    task = MulticlassClassificationTask(
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 270, in <module>
    main()
```
```
def __init__(self, nb_inputs, *args, **kwargs):
    self._nb_inputs = nb_inputs
    super().__init__(self, *args, **kwargs)
    self._softmax = softmax()
```
The above yields the following trace
```
Exception has occurred: TypeError
__init__() got multiple values for argument 'hidden_size'
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/task/reconstruction.py", line 179, in __init__
    super().__init__(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 206, in main
    task = MulticlassClassificationTask(
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 270, in <module>
    main()
```

Not that I expected a different outcome but I tested it.
```
def __init__(self, nb_inputs, *args, **kwargs):
    self._nb_inputs = nb_inputs
    super(Task, self).__init__(self, *args, **kwargs)
    self._softmax = softmax()
```
```
Exception has occurred: TypeError
__init__() got an unexpected keyword argument 'hidden_size'
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/task/reconstruction.py", line 179, in __init__
    super(Task,self).__init__(self, *args, **kwargs)
  File "/lustre/hpc/icecube/qgf305/workspace/graphnet/src/graphnet/models/config.py", line 242, in wrapper
    ret = init_fn(self, *args, **kwargs)
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 206, in main
    task = MulticlassClassificationTask(
  File "/groups/icecube/qgf305/workspace/analyses/multi_classification_on_stop_and_track_muons/modelling/train_multiclass_classification_model.py", line 270, in <module>
    main()
```


The code lives [here](https://github.com/MortenHolmRep/graphnet/tree/multiclass)
The loss function [here](https://github.com/MortenHolmRep/graphnet/blob/multiclass/src/graphnet/training/loss_functions.py)
The task [here](https://github.com/MortenHolmRep/graphnet/blob/multiclass/src/graphnet/models/task/reconstruction.py)